api/server.go

```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
+			return
+		}
+		siafundInput += sfo.Value
+	}
+
+	if siacoinInput.IsZero() && siafundInput == 0 {
+		jc.Error(errors.New("no inputs provided"), http.StatusBadRequest)
+	}
+
+	fee := s.cm.RecommendedFee().Mul64(2000) // use a const for simplicity
+
+	var sent bool
+	var locked []types.Hash256
+	defer func() {
+		if sent {
+			return
+		}
+		s.wm.Release(locked)
+	}()
+
+	sces, basis, siacoinChange, err := s.wm.SelectSiacoinElements(walletID, siacoinInput.Add(fee), false)
+	if err != nil {
+		jc.Error(fmt.Errorf("failed to select siacoin elements: %w", err), http.StatusInternalServerError)
+		return
+	}
+	for _, sce := range sces {
+		locked = append(locked, types.Hash256(sce.ID))
+	}
+
+	if !siacoinChange.IsZero() {
+		wcr.Siacoins = append(wcr.Siacoins, types.SiacoinOutput{
+			Value:   siacoinChange,
+			Address: wcr.ChangeAddress,
+		})
+	}
+
+	sfes, _, siafundChange, err := s.wm.SelectSiafundElements(walletID, siafundInput)
+	if err != nil {
+		jc.Error(fmt.Errorf("failed to select siafund elements: %w", err), http.StatusInternalServerError)
+		return
+	}
+	for _, sfe := range sfes {
+		locked = append(locked, types.Hash256(sfe.ID))
+	}
+
+	if siafundChange > 0 {
+		wcr.Siafunds = append(wcr.Siafunds, types.SiafundOutput{
+			Value:   siafundChange,
+			Address: wcr.ChangeAddress,
+		})
+	}
+
+	knownAddresses := make(map[types.Address]wallet.Address)
+	getAddress := func(addr types.Address) (wallet.Address, error) {
+		if a, ok := knownAddresses[addr]; ok {
+			return a, nil
+		}
+		a, err := s.wm.WalletAddress(walletID, addr)
+
+		if err != nil {
+			return wallet.Address{}, err
+		}
+		knownAddresses[addr] = a
+		return a, nil
+	}
+
+	resp := WalletConstructResponse{
+		Basis:        basis,
+		EstimatedFee: fee,
+	}
+
+	txn := types.Transaction{
+		MinerFees:      []types.Currency{fee},
+		SiacoinInputs:  make([]types.SiacoinInput, 0, len(sces)),
+		SiacoinOutputs: wcr.Siacoins,
+		SiafundInputs:  make([]types.SiafundInput, 0, len(sfes)),
+		SiafundOutputs: wcr.Siafunds,
+	}
+
+	for _, sce := range sces {
+		addr, err := getAddress(sce.SiacoinOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
+			return
+		}
+
+		sci := types.SiacoinInput{
+			ParentID: sce.ID,
+		}
+
+		if addr.SpendPolicy != nil {
+			// best effort to fill unlock conditions
+			uc, ok := addr.SpendPolicy.Type.(types.PolicyTypeUnlockConditions)
+			if !ok {
+				jc.Error(fmt.Errorf("address %q only unlock conditions are suppored in v1 transactions", addr.Address), http.StatusBadRequest)
+				return
+			}
+			sci.UnlockConditions = types.UnlockConditions(uc)
+		}
+
+		txn.SiacoinInputs = append(txn.SiacoinInputs, sci)
+		txn.Signatures = append(txn.Signatures, types.TransactionSignature{
+			ParentID: types.Hash256(sce.ID),
+			CoveredFields: types.CoveredFields{
+				WholeTransaction: true,
+			},
+		})
+	}
+
+	for _, sfe := range sfes {
```

Comment:
Not that it matters but bringing it up for consistency. One handler does siacoins first and then siafunds and the other one does siafunds first and then siacoins

Reply:
`UpdateV2TransactionSet` requires a V2Transaction as an argument, since we want to update the Siafund proofs to match the Siacoin proofs, we add the Siafund elements to the transaction first then check the basis and update the Siafund proofs.

Reply:
UpdateV2TransactionSet requires a V2Transaction as an argument, since we want to update the Siafund proofs to match the Siacoin proofs, we add the Siafund elements to the transaction first then check the basis and update the Siafund proofs.

Reply:
There's no technical reason we couldn't do it the other way around, just felt weird.

Reply:
Added a comment on *why* the ordering is the way it is:
```go
// the siafund elements are added to the transaction first because `UpdateV2TransactionSet` takes
// a V2 transaction as an argument. The Siacoin basis is our target because the transaction is
// guaranteed to have a non-zero Siacoin basis while the Siafund basis will be zero when not 
// sending Siafunds.
```